File: framework/test2/validate/list2d/list2d.js
const {ListNode, NestedInteger, TreeNode} = require('soda/leetcode');
const {TestWork, Utils, Validators} = require('soda/unittest');
const {
    PriorityQueue,
    MinPriorityQueue,
    MaxPriorityQueue
} = require('@datastructures-js/priority-queue');

// step [1]: implement solution function
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupByLength = function(strs) {
    shuffle(strs);
    let group = new Map();
    for (let s of strs) {
        if (!group.has(s.length)) {
            group.set(s.length, []);
        }
        group.get(s.length).push(s);
    }
    let keys = [...group.keys()];
    shuffle(keys);
    return keys.map(x => group.get(x));
};

function shuffle(arr) {
    for (let i = arr.length; i > 0; --i) {
        let index = Math.floor(Math.random() * i);
        if (index != i-1) {
            [arr[index], arr[i-1]] = [arr[i-1], arr[index]];
        }
    }
}

// step [2]: setup function/return/arguments
const work = TestWork.create(groupByLength);
// work = TestWork.forStruct(STRUCT);
work.validator = Validators.forArray2d('string', false, false);
work.compareSerial = true;
console.log(work.run(Utils.fromStdin()));

File: framework/test2/validate/list2d/List2d.kt
import kotlin.math.*
import soda.kotlin.leetcode.*
import soda.kotlin.unittest.*

class Solution {
    fun groupByLength(strs: Array<String>): Array<Array<String>> {
        val group = mutableMapOf<Int, MutableList<String>>()
        strs.shuffle()
        for (s in strs) {
            group.getOrPut(s.length, { mutableListOf<String>() }) += s
        }
        val keys = group.keys.toTypedArray()
        keys.shuffle()
        return keys.map { group[it]!!.toTypedArray() }.toTypedArray()
    }
}

class List2d : (String) -> String {
    override fun invoke(text: String): String {
        // * setup with your method
        val work = GenericTestWork.create(Solution()::groupByLength)

        // * by method that has not return value
        // val work = GenericTestWork.createVoid(Solution.METHOD_WITHOUT_RETURN)
        // * by class of data struct
        // val work = GenericTestWork.forStruct(STRUCT::class)

        // * setup validator
        // work.validator = (T, T) -> Boolean
        work.validator = Validators.forArray2d<String>(false, false)
        work.compareSerial = true
        return work.run(text)
    }

}

fun main() {
    print(List2d()(Utils.fromStdin()))
}

